requestor.go

```diff
@@ -0,0 +1,256 @@
+	}, nodeID)
+}
+
+func newSignedQuorumFromRound(round, seq uint64, signers []NodeID, myNodeID NodeID) *signedQuorum {
+	return &signedQuorum{
+		round:   round,
+		seq:     seq,
+		signers: NodeIDs(signers).Remove(myNodeID),
+	}
+}
+
+type sender interface {
+	// Send sends a message to the given destination node
+	Send(msg *Message, destination NodeID)
+}
+
+// requestor fetches quorum rounds up to [highestObserved] from the network,
+// allowing up to [maxRoundWindow] concurrent requests to limit memory use.
+// Ensures all rounds/sequences are eventually received.
+type requestor struct {
+	epochLock *sync.Mutex
+
+	// highestSequenceRequested prevents duplicates and limits outstanding requests.
+	highestRequested uint64
+
+	// the requestor stops requesting once all sequences/rounds up to an including `highestObserved` have been received.
+	highestObserved *signedQuorum
+
+	// Handles timeouts and retries for missing sequences/rounds.
+	timeoutHandler *TimeoutHandler[uint64]
+
+	logger Logger
+
+	// maxRoundWindow is the maximum number of requests we can request past highestRequested.
+	maxRoundWindow uint64
+
+	sender sender
+
+	// requestIterator is an iterator over NodeIDs in order to request quorum rounds
+	requestIterator int
+
+	// replicateSeqs is set true if this requestor is for replicating sequences, and false if for rounds.
+	replicateSeqs bool
+}
+
+func newRequestor(logger Logger, start time.Time, lock *sync.Mutex, maxRoundWindow uint64, sender sender, replicateSeqs bool) *requestor {
+	r := &requestor{
+		logger:         logger,
+		epochLock:      lock,
+		maxRoundWindow: maxRoundWindow,
+		sender:         sender,
+		replicateSeqs:  replicateSeqs,
+	}
+	name := "seq-th"
+	if !replicateSeqs {
+		name = "round-th"
+	}
+	r.timeoutHandler = NewTimeoutHandler(logger, name, start, DefaultReplicationRequestTimeout, r.resendReplicationRequests, shouldRemoveFunc[uint64](shouldDelete))
+	return r
+}
+
+func (r *requestor) advanceTime(now time.Time) {
+	r.timeoutHandler.Tick(now)
+}
+
+func (r *requestor) resendReplicationRequests(missingIds []uint64) {
+	// we call this function in the timeout handler goroutine, so we need to
+	// ensure we don't have concurrent access to highestObserved
+	r.epochLock.Lock()
+	defer r.epochLock.Unlock()
+
+	nodes := r.highestObserved.signers
+	numNodes := len(nodes)
+	slices.Sort(missingIds)
+	segments := CompressSequences(missingIds)
+	for i, seqsOrRounds := range segments {
+		index := (i + r.requestIterator) % numNodes
+		r.sendRequestToNode(seqsOrRounds.Start, seqsOrRounds.End, nodes[index])
+	}
+
+	r.requestIterator++
+}
+
+// maybeSendMoreReplicationRequests checks if we need to send more replication requests given an observed quorum.
+// it limits the amount of outstanding requests to be at most [maxRoundWindow] ahead of [currentSeqOrRound].
+func (r *requestor) maybeSendMoreReplicationRequests(observed *signedQuorum, currentSeqOrRound uint64) {
+	observedSeqOrRound := r.getSeqOrRound(observed)
```

Comment:
can we put a comment that we don't mix rounds and sequences because we have a different and mutually exclusive instance for each?